EIPS/eip-draft-trie-penalty.md

```diff
@@ -0,0 +1,261 @@
+- `dest` means destination already exists, or is a `precompile`
+- `!dest` means destination does not exist and is not a `precompile`
+
+| Op | value,dest| value, !dest |!value, dest| !value, !dest|  
+| -- | ---------  | -- | --| -- |
+|CALL |  no change | no change| no change| `penalty`|
+|CALLCODE |  no change | no change| no change| `penalty`|
+|DELEGATECALL | N/A | N/A| no change| `penalty` |
+|STATICCALL | N/A | N/A| no change| `penalty` |
+
+Whether the rules of this EIP is to be applied for regular ether-sends in `transactions` is TBD. See the 'Backwards Compatibility'-section for some more discussion on that topic. 
+
+### Note on `SELFDESTRUCT`
+
+The `SELFDESTRUCT` opcode also triggers an account trie lookup of the `beneficiary`. However, due to the following reasons, it has been omitted from having a `penalty` since it already costs `5K` gas. 
+
+### Clarifications:
+
+- The `base` costs of any opcodes are not modified by the EIP.
+- The opcode `SELFBALANCE` is not modified by this EIP, regardless of whether the `self` address exists or not. 
+
+### Determining the `penalty`
+
+A transaction with `10M` gas can today cause ~`14K` trie lookups.
+
+- A `penalty` of `1000`would lower the number to ~`5800` lookups, `41%` of the original.
+- A `penalty` of `2000`would lower the number to ~`3700` lookups, `26%` of the original.
+- A `penalty` of `3000`would lower the number to ~`2700` lookups, `20%` of the original. 
+- A `penalty` of `4000`would lower the number to ~`2100` lookups, `15%` of the original. 
+
+There exists a roofing function for the `penalty`. Since the `penalty` is deducted from `gas`, that means that a malicious contract can always invoke a malicious relay to perform the trie lookup. 
+
+In such a scenario, the `malicious` would spend `~750` gas each call to `relay`, and would need to provide the `relay` with at least `700` gas to do a trie access.
+
+Thus, the effective `cost` would be on the order of `1500`. It can thus be argued that `penalty` above `~800` would not achieve better protection against state-inexistence attacks. 
+
+
+## Rationale
+
+With this scheme, we could continue to price these operations based on the 'normal' usage, but gain protection from attacks that try to maximize disk lookups/cache misses. 
+This EIP does not modify anything regarding storage trie accesses, which might be relevant for a future EIP. However, there are a few crucial differences. 
+
+
+1. Storage tries are typically small, and there's a high cost to populate a storage trie with sufficient density for it to be in the same league as the account trie. 
+2. If an attacker wants to use an existing large storage trie, e.g. some popular token, he would typically have to make a `CALL` to cause a lookup in that token -- something like `token.balanceOf(<inexistant-address>)`. 
+   That adds quite a lot of extra gas-impediments, as each `CALL` is another `700` gas, plus gas for arguments to the `CALL`. 
```

Comment:
Is this dependent on the assumption that there are no popular contracts (i.e. ones with an existing large storage trie) with a method that reads a large number of (caller-determined) storage entries? If so, it would be nice to state this assumption explicitly here. (e.g. a theoretical popular token contract with a `sumBalanceOf(address[])` method or something like that)

Also if I understood you correctly, I would remove "plus gas for arguments to the CALL", since the `CALL` opcode doesn't charge gas for input data and presumably the DoS-attacker would generate the cache-miss arguments on-chain rather than passing them as tx data. The point about the minimum of 700 gas per cache-miss is still valid though.

Reply:
Well, I guess it could be several different ways to do this. However, it's much eaiser to get a 'random' number as a single stack-op, (e.g. `gas`), than if you have to pre-fill a memory segment with 'random' data and pass it to another method. Every argument you send is another 32 byte memory expanded on the caller-side. 

One could experiment more in-depth about exactly what the costs would be, but I don't think it would be marginal.

> Also if I understood you correctly, I would remove "plus gas for arguments to the CALL", since the CALL opcode doesn't charge gas for input data

You are correct that it doesn't, however, I was referring to the fact that you need to get your input first to stack, then load into memory. If you do it over and over (as opposed to a one-off multi-parameter thing), then you need to flip some bits between calls, to not hit the same thing over and over again.